### ISSUE TITLE
exception when implementing an associated type using an associated type projection

### ISSUE THREAD

**Issue by epdtry** [GaloisInc/mir-verifier#7]
Short example:

```Rust
impl<T: OtherTrait> Trait for Foo<T> {
    type Assoc = <T as OtherTrait>::OtherAssoc;
    fn dummy() {}
}
```

The exception occurs when the associated type `Assoc` is implemented using associated type projection `T::OtherAssoc`.  `fn dummy` is required - the bug doesn't trigger on traits with no associated fns.

Runnable example:

```Rust
pub trait Iterator {
    type Item;
    fn dummy();
}

impl<T: Iterator> Iterator for Option<T> {
    type Item = <T as Iterator>::Item;
    fn dummy() {}
}

// Stub to avoid "Could not find cfg: f"
const ARG: i32 = 1;
fn f(arg: i32) {}
```

crux-mir output:

```
$ cabal new-run crux-mir -- --no-std-lib test.rs
TOP LEVEL EXCEPTION
projection [::Iterator::Item<_0>] with unknown translation
CallStack (from HasCallStack):
  error, called at src/Mir/GenericOps.hs:186:19 in mir-verifier-0.1.0.0-inplace:Mir.GenericOps
  abstractATs_Ty, called at src/Mir/GenericOps.hs:400:17 in mir-verifier-0.1.0.0-inplace:Mir.GenericOps
  abstractATs, called at src/Mir/Pass/AssociatedTypes.hs:81:32 in mir-verifier-0.1.0.0-inplace:Mir.Pass.AssociatedTypes
  passAbstractAssociated, called at src/Mir/Trans.hs:2525:18 in mir-verifier-0.1.0.0-inplace:Mir.Trans
  transCollection, called at src/Mir/Generate.hs:117:38 in mir-verifier-0.1.0.0-inplace:Mir.Generate
```

This is blocking use of the original `Index for [T]` impls from libcore, which look like this:

```Rust
impl<T, I> ops::Index<I> for [T]
    where I: SliceIndex<[T]>
{
    type Output = I::Output;
    /* fns... */
}
```

**Comment by sweirich** [comment]
This is fixed by d91a31b5aac0cf122c10458e73bbde7df1fd5368, which can now translate the test case above (test/conc_eval/traits/tyfam5.rs). 

However, I still have trouble including the `Index for [T]` impl above in mir-lib/src/slice.rs because of the following errors from rustc:
```
error[E0119]: conflicting implementations of trait `core::ops::Index<_>` for type `[_]`:
    --> mir-lib/src/slice.rs:2589:1
     |
2589 | / impl<T, I> core::ops::Index<I> for [T]
2590 | |     where I: SliceIndex<[T]>
2591 | | {
2592 | |     type Output = I::Output;
...    |
2597 | |     }
2598 | | }
     | |_^
     |
     = note: conflicting implementation in crate `core`:
             - impl<T, I> core::ops::Index<I> for [T]
               where I: core::slice::SliceIndex<[T]>;

error[E0210]: type parameter `T` must be used as the type parameter for some local type (e.g. `MyStruct<T>`)
    --> mir-lib/src/slice.rs:2589:1
     |
2589 | / impl<T, I> core::ops::Index<I> for [T]
2590 | |     where I: SliceIndex<[T]>
2591 | | {
2592 | |     type Output = I::Output;
...    |
2597 | |     }
2598 | | }
     | |_^ type parameter `T` must be used as the type parameter for some local type
     |
     = note: only traits defined in the current crate can be implemented for a type parameter

error: aborting due to 2 previous errors
```

**Comment by epdtry** [comment]
>However, I still have trouble including the Index for [T] impl above in mir-lib/src/slice.rs because of the following errors from rustc:

The only fix I know of for that is to make the library not only `#[no_std]` but also `#[no_core]`.  The problem is, without linking to `core`, the library must now provide dozens of `#[lang]` items itself, just as `core` does.  Doing that with the current setup, where each source file in `mir-lib` is compiled as a separate crate, would be unpleasant or maybe even impossible (circular dependencies between crates are not allowed, while circular dependencies between modules of `core` almost certainly exist in some places).

We would need to restructure mir-lib to build as a single crate, and also provide implementations of any lang items that it uses but doesn't define.  (Fortunately, since tests and client programs can still build against regular `core` and use mir-lib only for verification, we shouldn't also need to provide every lang item a client might want to use on top of that.)

I'm closing this issue since it sounds like the main problem (the exception on associated types) is now fixed.
